lib/restbase.js

```diff
@@ -173,79 +173,131 @@ RESTBase.prototype.defaultWebRequestHandler = function(req) {
     return P.resolve(preq(req));
 };
 
+/**
+ * Checks if the maximum recursion depth has been exceeded by the request.
+ * If yes, the 500 error is thrown, othervise this is a no-op
+ *
+ * @param {Object} req - a current request object
+ * @private
+ */
+RESTBase.prototype._checkMaxRecursionDepth = function(req) {
+    if (this._recursionDepth > this._priv.options.maxDepth) {
+        var parents = [];
+        var rb = this._parent;
+        while (rb) {
+            parents.push(rb._req);
+            rb = rb._parent;
+        }
+        throw new HTTPError({
+            status: 500,
+            body: {
+                type: 'request_recursion_depth_exceeded',
+                title: 'RESTBase request recursion depth exceeded.',
+                uri: req.uri,
+                method: req.method,
+                parents: parents,
+                depth: this._recursionDepth
+            }
+        });
+    }
+};
 
-// Process one request
-RESTBase.prototype.request = function(req) {
-    // Protect the sys api from direct access
-    // Could consider opening this up with a specific permission later.
-    if (this._recursionDepth === 0 &&
-            ((req.uri.params && req.uri.params.api === 'sys')
-             // TODO: Remove once params.api is reliable
-             || (req.uri.path.length > 1 && req.uri.path[1] === 'sys'))) {
-        return P.reject(new HTTPError({
+/**
+ * Protects /sys APIs from the direct access.
+ *
+ * @param {Object} match - a found request handler
+ * @private
+ */
+RESTBase.prototype._checkInternalApiRequest = function(match) {
+    if (this._recursionDepth === 0 && match && match.params && match.params.api === 'sys') {
```

Comment:
I'd be prudent here and still keep the `req.uri.path[1] === 'sys'` check as a fall-back in case `/{api:sys}` is not specified in the URI on the bases that it is not strictly needed to actually put it there, hence preventing RB users to accidentally exposing internal routes to the public.